exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(4).Info("Machine has infrastructure, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) ([]clusterv1.Machine, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Construct a set of names of infraMachines that already have a Machine.
+	infraRefNames := sets.Set[string]{}
+	for _, machine := range machines {
+		infraRef := machine.Spec.InfrastructureRef
+		infraRefNames.Insert(infraRef.Name)
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		// If infraMachine already has a Machine, skip it.
+		if infraRefNames.Has(infraMachine.GetName()) {
+			continue
+		}
+
+		// Otherwise create a new Machine for the infraMachine.
+		log.V(4).Info("Creating new Machine for infraMachine", "infraMachine", infraMachine.GetName(), "namespace", infraMachine.GetNamespace())
+		machine := getNewMachine(mp, infraMachine)
+		if err := r.Client.Create(ctx, machine); err != nil {
+			return nil, errors.Wrapf(err, "failed to create new Machine for infraMachine %q in namespace %q", infraMachine.GetName(), infraMachine.GetNamespace())
+		}
+
+		machines = append(machines, *machine)
+	}
+
+	return machines, nil
+}
+
+// ensureInfraMachineOnwerRefs sets the ownerReferences on the each infraMachine to its associated MachinePool Machine if it hasn't already been done.
+func (r *MachinePoolReconciler) ensureInfraMachineOnwerRefs(ctx context.Context, updatedMachines []clusterv1.Machine, infraMachines []unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+	infraMachineNameToMachine := make(map[string]clusterv1.Machine)
+	for _, machine := range updatedMachines {
+		infraRef := machine.Spec.InfrastructureRef
+		infraMachineNameToMachine[infraRef.Name] = machine
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		ownerRefs := infraMachine.GetOwnerReferences()
+
+		machine, ok := infraMachineNameToMachine[infraMachine.GetName()]
+		if !ok {
+			return errors.Errorf("failed to patch ownerRef for infraMachine %q because no Machine has an infraRef pointing to it", infraMachine.GetName())
+		}
+		machineRef := metav1.NewControllerRef(&machine, machine.GroupVersionKind())
+		if !util.HasOwnerRef(ownerRefs, *machineRef) {
+			log.V(2).Info("Setting ownerRef on infraMachine", "infraMachine", infraMachine.GetName(), "namespace", infraMachine.GetNamespace(), "machine", machine.GetName())
+			ownerRefs = util.EnsureOwnerRef(ownerRefs, *machineRef)
+			infraMachine.SetOwnerReferences(ownerRefs)
+			if err := r.Client.Update(ctx, infraMachine); err != nil {
```

Comment:
I'd have to double check, but I used the patch helper in the beginning and it didn't update the ownerRef and when I switched to Update, it worked.

Reply:
Please let's try to make the patchHelper work. We use it everywhere and just calling update here is not ideal. I would expect it to run into conflicts if other controllers are concurrently writing the machines (or in case optimistic locking is not used, I would expect it to overwrite changes of other controllers which is also not good)

Reply:
Update: I think we got this working based on a Slack chat